### PR TITLE
MI-198: python 3 compatibility; includes:

### PR DESCRIPTION
**Pull request by lbjay**
* using pip-compile to generate requirements
* adding black and flake8 rules and pre-commit hook
* python 3 updates, almost all done using the `2to3` command
* arrow objects now use `shift` instead of `replace`
* a bunch of black formatting changes (wish i'd done this separately)
* removing some loggly cruft that isn't used anymore

**note**: although I added the config files for black, flake8, etc. I did not introduce a `dev` version of the requirements. It's important for the virtualenv i'm using for development be python  3.4 and those tools don't have support for python's that old. I'm just using "global" installs of those tools.

### CODE REVIEW

**Comment by lbjay** [comment]
A gist showing just what the 2to3 changes were. https://gist.github.com/lbjay/76468f21edf2abea028d79f1f92db96c

 I admit I didn't look to closely at these. I'm not quite sure what the difference is between `for group in resp_data['people'].values():` and `for group in list(resp_data['people'].values()):`.

**Comment by lbjay** [comment]
I found a bug in the code using the arrow library. Don't review yet.

File: harvest_cli/ocua.py [hunk 1]
```diff
@@ -123,170 +188,241 @@ def useractions(start, end, wait, engage_host, user, password, output, queue_nam
...
     return rec
 
+
 def get_episode(action):
     cached_ep = r.get(action.mediapackageId)
     if cached_ep is not None:
         logger.debug("episode cache hit for %s", action.mediapackageId)
-        episode_data = json.loads(cached_ep)
-        episode_data['__from_cache'] = True
+        episode_data = json.loads(cached_ep.decode("utf-8"))
```
Comment (lbjay) [inline harvest_cli/ocua.py:327]: had to add the `.decode` here

File: harvest_cli/ocua.py [hunk 2]
```diff
@@ -298,116 +434,146 @@ def load_episodes(admin_host, engage_host, user, password, es_host, target_index
...
     bucket = get_or_create_bucket()
     bucket.put_object(Key=ts_key, Body=timestamp)
 
+
 def get_harvest_ts(ts_key):
     bucket = get_or_create_bucket()
     try:
         obj = bucket.Object(ts_key).get()
-        return obj['Body'].read()
+        return obj["Body"].read().decode("utf-8")
```
Comment (lbjay) [inline harvest_cli/ocua.py:574]: added `.decode` here too

**Comment by lbjay** [comment]
OK, PR updated. There were a couple of spots where i was treating a bytes as a string and python3 doesn't like that.

File: harvest_cli/__init__.py [hunk 3]
```diff
@@ -1,21 +1,24 @@
...
+__version__ = "0.1.0"
 
 logger = logging.getLogger(__name__)
 
+
 @click.group()
-@click.option('--log-level', default='info',
-              type=click.Choice(['info','debug','warn']))
+@click.option(
+    "--log-level", default="info", type=click.Choice(["info", "debug", "warn"])
```
Comment (nataliemona) [inline harvest_cli/__init__.py]: Magic comma

File: harvest_cli/dev.py [hunk 4]
```diff
@@ -54,27 +54,33 @@ def install_kopf():
...
-                            if 'elasticsearch' in x.name)
+        es_container = next(
+            x
+            for x in docker_client.containers.list()
+            if "elasticsearch" in x.name
+        )
         click.echo("Installing kopf plugin")
         exec_inst = docker_client.api.exec_create(
-            es_container.id, "bin/plugin install lmenezes/elasticsearch-kopf")
+            es_container.id, "bin/plugin install lmenezes/elasticsearch-kopf"
```
Comment (nataliemona) [inline harvest_cli/dev.py]: Magic comma

File: harvest_cli/export.py [hunk 5]
```diff
@@ -56,76 +54,70 @@ def viewing_options(es_host, term, year):
...
 
-    attendance_column = live and 'live_attendance' or 'vod_attendance'
-    duration_attr = live and 'live_duration' or 'duration'
+    attendance_column = live and "live_attendance" or "vod_attendance"
+    duration_attr = live and "live_duration" or "duration"
 
     writer = DictWriter(
-        sys.stdout,
-        fieldnames=['mpid', 'huid', attendance_column]
+        sys.stdout, fieldnames=["mpid", "huid", attendance_column]
```
Comment (nataliemona) [inline harvest_cli/export.py]: Magic comma

File: harvest_cli/ocua.py [hunk 6]
```diff
@@ -11,83 +11,148 @@
 import re
 
 import pyhorn
+
 # force all calls to the episode search endpoint to use includeDeleted=true
-pyhorn.endpoints.search.SearchEndpoint._kwarg_map['episode']['includeDeleted'] = True
+pyhorn.endpoints.search.SearchEndpoint._kwarg_map["episode"][
+    "includeDeleted"
+] = True
```
Comment (nataliemona) [inline harvest_cli/ocua.py]: Weird black formatting... ZIP has a few of these too... I don't know

File: harvest_cli/ocua.py [hunk 7]
```diff
@@ -11,83 +11,148 @@
...
-@click.option('--update-last-ts/--no-update-last-ts', is_flag=True,
-              default=True, help="Update the last harvested timestamp; "
-                   "use --no-update-last-ts if backfilling data")
-def useractions(start, end, wait, engage_host, user, password, output, queue_name,
-                batch_size, interval, disable_start_end_span_check,
-                update_last_ts):
+@click.option("-s", "--start", help="YYYYMMDDHHmmss")
+@click.option("-e", "--end", help="YYYYMMDDHHmmss; default=now")
+@click.option(
+    "-w", "--wait", default=1, help="Seconds to wait between batch requests"
```
Comment (nataliemona) [inline harvest_cli/ocua.py]: Magic comma

File: harvest_cli/ocua.py [hunk 8]
```diff
@@ -11,83 +11,148 @@
...
+)
+@click.option(
+    "-q",
+    "--queue-name",
+    envvar="SQS_QUEUE_NAME",
+    help="SQS queue name",
+    required=True,
+)
+@click.option(
+    "-b", "--batch-size", default=1000, help="number of actions per request"
```
Comment (nataliemona) [inline harvest_cli/ocua.py]: Magic comma

File: harvest_cli/zoom.py [hunk 9]
```diff
@@ -7,40 +7,59 @@
...
+    default="index",
+)
+@click.option(
+    "--es_host",
+    envvar="ES_HOST",
+    default="localhost:9200",
+    help="Elasticsearch host:port; defaults to $ES_HOST",
+)
+@click.option(
+    "--key", envvar="ZOOM_KEY", help="zoom api key; defaults to $ZOOM_KEY"
```
Comment (nataliemona) [inline harvest_cli/zoom.py]: Magic comma

File: harvest_cli/zoom.py [hunk 10]
```diff
@@ -99,7 +119,9 @@ class ZoomApiException(Exception):
     pass
 
 
-def fetch_records(report_url, params, listkey, countkey='total_records', wait=1):
+def fetch_records(
+    report_url, params, listkey, countkey="total_records", wait=1
```
Comment (nataliemona) [inline harvest_cli/zoom.py:125]: Magic comma

File: harvest_cli/export.py [hunk 11]
```diff
@@ -56,76 +54,70 @@ def viewing_options(es_host, term, year):
...
         res = s.execute().to_dict()
-        for huid_bucket in res['aggregations']['huid']['buckets']:
-            huid = huid_bucket['key']
-            row = {'mpid': mp.mpid, 'huid': huid}
-            for event_type, stats in huid_bucket['paella_events']['buckets'].items():
-                row[event_type] = stats['doc_count']
+        for huid_bucket in res["aggregations"]["huid"]["buckets"]:
+            huid = huid_bucket["key"]
+            row = {"mpid": mp.mpid, "huid": huid}
+            for event_type, stats in list(
```
Comment (nataliemona) [inline harvest_cli/export.py]: Remove list

File: harvest_cli/zoom.py [hunk 12]
```diff
@@ -111,43 +133,46 @@ def fetch_records(report_url, params, listkey, countkey='total_records', wait=1)
         r.raise_for_status()
         response = r.json()
 
-        if 'error' in response.keys():
-            error_code = response['error']['code']
-            error_message = response['error']['message']
+        if "error" in list(response.keys()):
```
Comment (nataliemona) [inline harvest_cli/zoom.py]: Remove list

**Review by nataliemona — APPROVED** [review]

**Pull request merged**